### PR TITLE
[Feature] Add bias in node decoder

### PR DESCRIPTION
**Pull request by RonaldBXu**
*Issue #, if available:*

https://github.com/awslabs/graphstorm/issues/335

*Description of changes:*

Adding bias in node decoder and unit tests.

Regression Test W/ Bias:
```
Dataset		Test Performance	Validation Performance	Total Time	Epoch Time/Seconds
OGB-MAG-NC	Acc: 0.4023		Acc: 0.4149		0:24:55		68.511
OGB-PROD-NC	Acc: 0.7178		Acc: 0.9117		0:47:22		110.806
```

Regression Test W/O Bias:
```
Dataset		Test Performance	Validation Performance	Total Time	Epoch Time/Seconds
OGB-MAG-NC	Acc: 0.4067		Acc: 0.4113		0:23:56		63.323
OGB-PROD-NC	Acc: 0.7250		Acc: 0.9138		0:46:08		104.001
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/unit-tests/test_decoder.py [hunk 1]
```diff
@@ -981,16 +982,60 @@ def test_MLPEFeatEdgeDecoder(h_dim, feat_dim, out_dim, num_ffn_layers):
...
     assert decoder.out_dims == 1
 
-    decoder = EntityRegression(h_dim=in_dim, out_dim=out_dim)
+    decoder = EntityRegression(h_dim=in_dim, out_dim=out_dim, bias=True)
+    in_tensor = th.ones((1,in_dim))
+    with th.no_grad():
+        decoder.eval()
+        decoder.decoder.data = th.ones((in_dim, out_dim))
+        out = decoder.predict(in_tensor)
+    assert th.all(out == in_dim)
```
Comment (zhjwy9343) [inline tests/unit-tests/test_decoder.py:1011]: why use `in_dim` to measure output? Should it be `out_dim`?

File: tests/unit-tests/test_decoder.py [hunk 2]
```diff
@@ -981,16 +982,60 @@ def test_MLPEFeatEdgeDecoder(h_dim, feat_dim, out_dim, num_ffn_layers):
         prediction = decoder.predict(g, encoder_feat, efeat)
         pred = out.argmax(dim=1)
         assert_almost_equal(prediction.cpu().numpy(), pred.cpu().numpy())
+
+def check_dropout_train(decoder, in_dim):
```
Comment (zhjwy9343) [inline tests/unit-tests/test_decoder.py:986]: Could you add comments of this function to explain its purpose and logics?

File: tests/unit-tests/test_decoder.py [hunk 3]
```diff
@@ -1010,6 +1055,8 @@ def test_EntityRegression(in_dim, out_dim):
     test_EntityRegression(8, 1)
     test_EntityRegression(8, 8)
 
+    test_EntityClassifier(64, 8)
```
Comment (zhjwy9343) [inline tests/unit-tests/test_decoder.py:1058]: - This PR is for decoder bias, but there is no test case for it.
- The `test_EntityClassifier` has multiple input combination, should we test more combination here?

File: tests/unit-tests/test_decoder.py [hunk 4]
```diff
@@ -981,16 +982,60 @@ def test_MLPEFeatEdgeDecoder(h_dim, feat_dim, out_dim, num_ffn_layers):
...
         pred = out.argmax(dim=1)
         assert_almost_equal(prediction.cpu().numpy(), pred.cpu().numpy())
+
+def check_dropout_train(decoder, in_dim):
+    ones = th.ones(in_dim)
+    decoder.train()
+    output = decoder(ones)
+    assert th.any(output == 0)
+
+def check_dropout_eval(decoder, in_dim):
```
Comment (zhjwy9343) [inline tests/unit-tests/test_decoder.py:992]: ditto

**Review by zhjwy9343 — COMMENTED** [review]
This PR added decoder bias to node decoder only. Any reason why no add to edge decoders?

File: python/graphstorm/config/argument.py [hunk 5]
```diff
@@ -1749,6 +1752,18 @@ def log_report_frequency(self):
...
 
     ###################### Model training related ######################
+    @property
+    def decoder_bias(self):
+        """ Decoder bias. decoder_bias must be a boolean. Default is False.
+        """
+        # pylint: disable=no-member
+        if hasattr(self, "_decoder_bias"):
+            assert self._decoder_bias in [True, False], \
+                "use_early_stop should be in [True, False]"
```
Comment (classicsong) [inline python/graphstorm/config/argument.py:1762]:
```suggestion
                "decoder_bias should be in [True, False]"
```

File: python/graphstorm/model/node_decoder.py [hunk 6]
```diff
@@ -38,20 +38,24 @@ class EntityClassifier(GSLayer):
     norm: str
         Normalization methods. Not used, but reserved for complex node classifier
         implementation. Default: None.
+    bias: bool
+        Whether the node decoder has bias. Default: False.
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:42]:
```suggestion
        Whether the node decoder uses a bias parameter. Default: False.
```

File: python/graphstorm/model/node_decoder.py [hunk 7]
```diff
@@ -38,20 +38,24 @@ class EntityClassifier(GSLayer):
...
+                 norm=None,
+                 bias=False):
         super(EntityClassifier, self).__init__()
         self._in_dim = in_dim
         self._num_classes = num_classes
         self._multilabel = multilabel
         self._dropout = dropout
         # TODO(xiangsx): The norm is not used here.
         self._norm = norm
+        self._bias = bias
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:58]:
```suggestion
        self._use_bias = bias
```
It is misleading if we use _bias as a flag and bias as a nn.Parameter.

File: python/graphstorm/model/node_decoder.py [hunk 8]
```diff
@@ -77,9 +83,15 @@ def forward(self, inputs):
 
         Returns
         -------
-        Tensor: the prediction logits.
+        output: Tensor of the prediction logits.
         '''
-        return th.matmul(inputs, self.decoder)
+        out = th.matmul(inputs, self.decoder)
+        if self._bias:
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:89]: ditto

File: python/graphstorm/model/node_decoder.py [hunk 9]
```diff
@@ -77,9 +83,15 @@ def forward(self, inputs):
 
         Returns
         -------
-        Tensor: the prediction logits.
+        output: Tensor of the prediction logits.
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:86]:
```suggestion
        out: Tensor of the prediction logits.
```

File: python/graphstorm/model/node_decoder.py [hunk 10]
```diff
@@ -91,10 +103,13 @@ def predict(self, inputs):
 
         Returns
         --------
-        Tensor: argmax of the prediction results, or the maximum of the prediction results
+        out: Tensor of argmax of the prediction results, or the maximum of the prediction results
         if ``multilabel`` is ``True``.
         """
-        logits = th.matmul(inputs, self.decoder)
+        logits = th.matmul(inputs, self.decoder) 
+        if self._bias:
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:110]: ditto

File: python/graphstorm/model/node_decoder.py [hunk 11]
```diff
@@ -115,9 +130,12 @@ def predict_proba(self, inputs):
 
         Returns
         -------
-        Tensor: normalized prediction results.
+        out: Tensor of normalized prediction results.
         """
         logits = th.matmul(inputs, self.decoder)
+        if self._bias:
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:136]: ditto

File: python/graphstorm/model/node_decoder.py [hunk 12]
```diff
@@ -153,24 +171,30 @@ class EntityRegression(GSLayer):
     norm: str, optional
         Normalization methods. Not used, but reserved for complex node regression
         implementation. Default: None.
+    bias: bool
+        Whether the node decoder has bias. Default: False.
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:175]: ditto

File: tests/unit-tests/test_decoder.py [hunk 13]
```diff
@@ -981,16 +982,60 @@ def test_MLPEFeatEdgeDecoder(h_dim, feat_dim, out_dim, num_ffn_layers):
...
+
 @pytest.mark.parametrize("in_dim", [16, 64])
 @pytest.mark.parametrize("out_dim", [1, 8])
 def test_EntityRegression(in_dim, out_dim):
     decoder = EntityRegression(h_dim=in_dim)
     assert decoder.in_dims == in_dim
     assert decoder.out_dims == 1
 
-    decoder = EntityRegression(h_dim=in_dim, out_dim=out_dim)
+    decoder = EntityRegression(h_dim=in_dim, out_dim=out_dim, bias=True)
```
Comment (classicsong) [inline tests/unit-tests/test_decoder.py:1005]: You should test both bias = True and bias = False.

File: tests/unit-tests/test_decoder.py [hunk 14]
```diff
@@ -981,16 +982,60 @@ def test_MLPEFeatEdgeDecoder(h_dim, feat_dim, out_dim, num_ffn_layers):
...
+    in_tensor = th.ones((1,in_dim))
+    with th.no_grad():
+        decoder.eval()
+        decoder.decoder.data = th.ones((in_dim, out_dim))
+        out = decoder.predict(in_tensor)
```
Comment (classicsong) [inline tests/unit-tests/test_decoder.py:1010]: Can you also test the case when decoder.bias is not a zero tensor?
You can hack the decoder.bias for the test.

File: tests/unit-tests/test_decoder.py [hunk 15]
```diff
@@ -981,16 +982,60 @@ def test_MLPEFeatEdgeDecoder(h_dim, feat_dim, out_dim, num_ffn_layers):
...
     assert decoder.in_dims == in_dim
     assert decoder.out_dims == out_dim
+    check_dropout_train(decoder, in_dim)
+    check_dropout_eval(decoder, in_dim)
+
+@pytest.mark.parametrize("in_dim", [16, 64])
+@pytest.mark.parametrize("num_classes", [4, 8])
+def test_EntityClassifier(in_dim, num_classes):
+
+    decoder = EntityClassifier(in_dim=in_dim, num_classes=num_classes, multilabel=False, bias=True)
```
Comment (classicsong) [inline tests/unit-tests/test_decoder.py:1023]: ditto

**Comment by RonaldBXu** [comment]
@zhjwy9343 I think Xiang said to just do node decoder bias for now and edge decoder bias in a later pr.

File: python/graphstorm/model/node_decoder.py [hunk 16]
```diff
@@ -55,7 +55,7 @@ def __init__(self,
         self._dropout = dropout
         # TODO(xiangsx): The norm is not used here.
         self._norm = norm
-        self._bias = bias
+        self._use_bias = bias
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:58]:
```suggestion
        self._use_bias = use_bias
```

File: python/graphstorm/model/node_decoder.py [hunk 17]
```diff
@@ -186,14 +186,14 @@ def __init__(self,
...
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py:189]:
```suggestion
        self._use_bias = use_bias
```

File: python/graphstorm/model/node_decoder.py (hunk 16 above)
Comment (RonaldBXu) [reply python/graphstorm/model/node_decoder.py:58]: i thought i already did this?

File: python/graphstorm/model/node_decoder.py (hunk 17 above)
Comment (RonaldBXu) [reply python/graphstorm/model/node_decoder.py:189]: same as above

File: python/graphstorm/config/argument.py [hunk 18]
```diff
@@ -1749,6 +1750,18 @@ def log_report_frequency(self):
...
+    @property
+    def decoder_bias(self):
+        """ Decoder bias. decoder_bias must be a boolean. Default is False.
+        """
+        # pylint: disable=no-member
+        if hasattr(self, "_decoder_bias"):
+            assert self._decoder_bias in [True, False], \
+                "decoder_bias should be in [True, False]"
+            return self._decoder_bias
+        # By default, decoder bias is False
```
Comment (jalencato) [inline python/graphstorm/config/argument.py:1762]: Shall we comment here the bias here is only for node/edge?

File: python/graphstorm/model/node_decoder.py [hunk 19]
```diff
@@ -77,9 +83,15 @@ def forward(self, inputs):
...
         -------
-        Tensor: the prediction logits.
+        output: Tensor of the prediction logits.
         '''
-        return th.matmul(inputs, self.decoder)
+        out = th.matmul(inputs, self.decoder)
+        if self._bias:
+            out = out + self.bias
+
+        out = self.dropout(out)
```
[the remark this answers is not in the document]
Comment (jalencato) [reply python/graphstorm/model/node_decoder.py:92]: Agree. We better not to make dropout apply to all the cases.

**Review by jalencato — COMMENTED** [review]
Could you add some performance number in the PR description for w/wo bias?

File: python/graphstorm/model/node_decoder.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (RonaldBXu) [reply python/graphstorm/model/node_decoder.py:92]: Ok, it seems people on the internet also agree, so I will remove it.

File: tests/unit-tests/test_decoder.py (hunk 1 above)
Comment (RonaldBXu) [reply tests/unit-tests/test_decoder.py:999]: I will add this assertion:

assert out.shape == (1, out_dim)

but i think the actual values of out are the size of the input? since my input are all ones.

File: tests/unit-tests/test_decoder.py [hunk 20]
```diff
@@ -981,17 +982,62 @@ def test_MLPEFeatEdgeDecoder(h_dim, feat_dim, out_dim, num_ffn_layers):
...
+    decoder = EntityClassifier(in_dim=in_dim, num_classes=num_classes, multilabel=False, use_bias=True)
+    assert decoder.in_dims == in_dim
+    assert decoder.out_dims == num_classes
+
+    decoder.eval()
+
+    in_tensor = th.ones(1, in_dim)
+    with th.no_grad():
+        th.nn.init.eye_(decoder.decoder)
+        decoder.decoder.data[0][0] += 1
```
Comment (zhjwy9343) [inline tests/unit-tests/test_decoder.py:1032]: What is the purpose of this assignment?
It seems if the `in_tensor` is `th.ones(1, in_dim)` and using `.init.eye_`, the results of `decoder.predict(in_tensor)` will always be 0 no matter `+=1` or not.

In addition, it is very hard to see the logic of this test case. My guess is that by adding a positive value in the `.data[0][N]` dimension, the predicted class will be the `N`. If this is what you want to test `EntityClassifier`, it is better to use a variable to control, showing this logic explicitly. And adding some comment lines will help too.

File: tests/unit-tests/test_decoder.py [hunk 21]
```diff
@@ -981,17 +982,62 @@ def test_MLPEFeatEdgeDecoder(h_dim, feat_dim, out_dim, num_ffn_layers):
...
+        out = decoder.predict(in_tensor)
+    assert out == 0
+
+    with th.no_grad():
+        decoder.bias.data[3] += 2
```
Comment (zhjwy9343) [inline tests/unit-tests/test_decoder.py:1037]: ditto

File: python/graphstorm/model/node_decoder.py [hunk 22]
```diff
@@ -89,8 +90,6 @@ def forward(self, inputs):
         if self._use_bias:
             out = out + self.bias
 
-        out = self.dropout(out)
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py]: Why remove the dropout?
We can move it before L88 (th.matmul).

File: python/graphstorm/model/node_decoder.py [hunk 23]
```diff
@@ -218,8 +217,6 @@ def forward(self, inputs):
...
```
Comment (classicsong) [inline python/graphstorm/model/node_decoder.py]: ditto

**Review by classicsong — APPROVED** [review]
LGTM.

**Review by zhjwy9343 — APPROVED** [review]
LGTM now
